lxd/project/permissions.go

```diff
@@ -1514,3 +1514,113 @@ func AllowSnapshotCreation(p *api.Project) error {
 
 	return nil
 }
+
+// GetRestrictedClusterGroups returns a slice of restricted cluster groups for the given project.
+func GetRestrictedClusterGroups(p *api.Project) []string {
+	return shared.SplitNTrimSpace(p.Config["restricted.cluster.groups"], ",", -1, true)
+}
+
+// AllowClusterMember returns nil if the given project is allowed to use the cluster member.
+func AllowClusterMember(p *api.Project, member *db.NodeInfo) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	if shared.IsTrue(p.Config["restricted"]) && len(clusterGroupsAllowed) > 0 {
+		for _, memberGroupName := range member.Groups {
+			if shared.StringInSlice(memberGroupName, clusterGroupsAllowed) {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("Project isn't allowed to use this cluster member: %q", member.Name)
+	}
+
+	return nil
+}
+
+// AllowClusterGroup returns nil if the given project is allowed to use the cluster group.
+func AllowClusterGroup(p *api.Project, groupName string) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	// Skip the check if the project is not restricted
+	if shared.IsFalseOrEmpty(p.Config["restricted"]) {
+		return nil
+	}
+
+	if len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(groupName, clusterGroupsAllowed) {
+		return fmt.Errorf("Project isn't allowed to use this cluster group: %q", groupName)
+	}
+
+	return nil
+}
+
+// CheckTargetMember checks if the given member is allowed for the project.
+// If nil it also returns the resolved target member node information.
```

Comment:
This comment describes the internal implementation and requires the reader to understand it, as it doesn't say what is `nil` in this case (I'm assuming you mean the return value of `AllowClusterMember`).

Ideally the reader of the docblock shouldn't need to see the internal implementation of a function to understand its return values.

I suggest something like this:

> // CheckTargetMember checks if the given targetMemberName is present in allMembers and is allowed for the project.
> // If member is allowed it returns the resolved target member information.

Reply:
Changed as it is much easier to deal with the functions if the text matches the variable name. Also modified the doc strings for the other functions were we added the `{...}Name` suffix.